syntax="proto3";

package descarteslabs.workflows;

import "google/protobuf/descriptor.proto";


extend google.protobuf.FieldOptions {
  string mimetype = 3226;
}


// NOTE: We use a 'not_' naming convention for boolean fields
// whose default values are True. The default value of a boolean field in
// proto3 is False; this convention allows us properly handle boolean fields with
// a default value of True. Applications presenting these fields to
// users should do so without the 'not_' prefix and with a default of True.
// Mimetypes support passing these parameters without the 'not_' prefix,
// in which case the value is inverted.

message Format {
  // TODO: add oneof for these once we can handle it better on the backend
  Pyarrow pyarrow = 1 [(mimetype) = "application/vnd.pyarrow"];
  JSON json = 2 [(mimetype) = "application/json"];
  GeoJSON geojson = 3 [(mimetype) = "application/geo+json"];
  CSV csv = 4 [(mimetype) = "text/csv"];
  PNG png = 5 [(mimetype) = "image/png"];
  Geotiff geotiff = 6 [(mimetype) = "image/tiff"];
  MsgPack msgpack = 7 [(mimetype) = "application/msgpack"];

  bool has_pyarrow = 20;
  bool has_json = 21;
  bool has_geojson = 22;
  bool has_csv = 23;
  bool has_png = 24;
  bool has_geotiff = 25;
  bool has_msgpack = 26;
}


message Pyarrow {
    enum PyarrowCompression {
        PYARROWCOMPRESSION_UNSPECIFIED = 0;
        PYARROWCOMPRESSION_LZ4 = 1; // default
        PYARROWCOMPRESSION_BROTLI = 2;
    }
    PyarrowCompression compression = 1;
}

message JSON {}

message GeoJSON {}

message CSV {}

message PNG {}



message Geotiff {
    bool not_overviews = 1;
    enum GeotiffCompression {
        GEOTIFFCOMPRESSION_UNSPECIFIED = 0;
        GEOTIFFCOMPRESSION_LZW = 1; // default
        GEOTIFFCOMPRESSION_NONE = 2;
        GEOTIFFCOMPRESSION_JPEG = 3;
    }
    GeotiffCompression compression = 2;
    enum GeotiffOverviewResampler {
      GEOTIFFOVERVIEWRESAMPLER_UNSPECIFIED = 0;
      GEOTIFFOVERVIEWRESAMPLER_NEAREST = 1; // default
      GEOTIFFOVERVIEWRESAMPLER_AVERAGE = 2;
      GEOTIFFOVERVIEWRESAMPLER_BILINEAR = 3;
      GEOTIFFOVERVIEWRESAMPLER_CUBIC = 4;
      GEOTIFFOVERVIEWRESAMPLER_MODE = 5;
    }
    GeotiffOverviewResampler overview_resampler = 3;
}

message MsgPack {}
